src/main/java/org/jabref/model/database/BibDatabaseContext.java

```diff
@@ -154,40 +153,45 @@ public boolean isStudy() {
      *     <li>user-specific metadata directory</li>
      *     <li>general metadata directory</li>
      *     <li>BIB file directory (if configured in the preferences AND none of the two above directories are configured)</li>
-     *     <li>preferences directory (if .bib file directory should not be used according to the preferences)</li>
+     *     <li>preferences directory (if .bib file directory should not be used according to the (global) preferences)</li>
      * </ol>
      *
      * @param preferences The fileDirectory preferences
      */
     public List<Path> getFileDirectories(FilePreferences preferences) {
-        List<Path> fileDirs = new ArrayList<>();
+        List<Path> fileDirs = new ArrayList<>(3);
 
-        // 1. Metadata user-specific directory
-        metaData.getUserFileDirectory(preferences.getUserAndHost())
-                .ifPresent(userFileDirectory -> fileDirs.add(getFileDirectoryPath(userFileDirectory)));
+        Optional<Path> userFileDirectory = metaData.getUserFileDirectory(preferences.getUserAndHost()).map(dir -> getFileDirectoryPath(dir));
+        userFileDirectory.ifPresent(fileDirs::add);
 
-        // 2. Metadata general directory
-        metaData.getDefaultFileDirectory()
-                .ifPresent(metaDataDirectory -> fileDirs.add(getFileDirectoryPath(metaDataDirectory)));
+        Optional<Path> generalFileDirectory = metaData.getDefaultFileDirectory().map(dir -> getFileDirectoryPath(dir));
+        generalFileDirectory.ifPresent(fileDirs::add);
 
-        // 3. BIB file directory or main file directory
-        // fileDirs.isEmpty in the case, 1) no user-specific file directory and 2) no general file directory is set
-        // (in the metadata of the bib file)
-        if (fileDirs.isEmpty() && preferences.shouldStoreFilesRelativeToBibFile()) {
+        // fileDirs.isEmpty() is true after these two if there are no directories set in the BIB file itself:
+        //   1) no user-specific file directory set (in the metadata of the bib file) and
+        //   2) no general file directory is set (in the metadata of the bib file)
+
+        // BIB file directory or main file directory (according to (global) preferences)
+        if (preferences.shouldStoreFilesRelativeToBibFile()) {
             getDatabasePath().ifPresent(dbPath -> {
                 Path parentPath = dbPath.getParent();
                 if (parentPath == null) {
                     parentPath = Path.of(System.getProperty("user.dir"));
+                    LOGGER.warn("Parent path of database file {} is null. Falling back to {}.", dbPath, parentPath);
                 }
                 Objects.requireNonNull(parentPath, "BibTeX database parent path is null");
-                fileDirs.add(parentPath);
+                Path absolutePath = parentPath.toAbsolutePath();
+                if (!fileDirs.contains(absolutePath)) {
+                    fileDirs.add(absolutePath);
+                }
```

Comment:
I believe we can make a set out of this...

Reply:
Not useful, we need to keep the order and we only have like 3 possible values

Reply:
sortedset

Reply:
> sortedset

We need to keep the order.

Sequenced Collection is the "right" data type. It checks for containment at each insert. With my approach, only one check needs to be made.

Moreover, we have at most two elements to check. We use an ArrayList where the "forEach" is faster than with a LinkedList.

Changing the return type to a sequencedcollection results in much code changes. Using it internally and then converting to a list results in more CPU cycles than "just" checking two elements.

I can rewrite to return SequencedCollection, write an ADR, add Java comments, link to this text or just leave it.

What should I do?

Reply:
OK, I just switched to `LinkedHashSet`. This also ensures that the second path is checked for uniqueness.

Reply:
Was just an idea, the code just smelled like a set. no strong opinion on this.